[PEP 646] Allow cleanly substituting any tuple type for a TypeVarTuple

pradeep90 opened the pull request:
Our specification implicitly treated it as legal to replace a `TypeVarTuple` with a tuple such as `Tuple[int, str]`. However, we did not mention the natural implications for `*args` or `Callable`. We also did not address the behavior of unbounded tuples.

This led to artificial restrictions (such as "type variable tuples must be of known length").

This PR specifies that, wherever a `Ts` is used, we can legally replace it with `Tuple[int, str]` or `Tuple[int, ...]` or `Tuple[int, *Ts, str]`. Concretely, the following are valid:
+ `*args: *Tuple[int, *Ts, str]`, etc.
+ `Callable[[int, *Ts, str], None]`, etc.
+ `Tuple[int, *Tuple[str, ...]]`, etc.

This replaces #2125 .

cc @mrahtz @gvanrossum

gvanrossum commented:
I'll review this once @mrahtz has approved it.

mrahtz commented on pep-0646.rst:
```diff
@@ -414,6 +354,133 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+    process_batch_channels(z)  # Error: Expected Channels.
+
+
+We can also pass a ``Tuple[int, ...]`` wherever a ``Tuple[*Ts]`` is
```
I found this first sentence a bit confusing - is it saying that we can pass a ``Tuple[int, ...]`` wherever a ``*Ts`` appears in the argument list?

mrahtz commented on pep-0646.rst:
```diff
@@ -428,13 +495,59 @@ individual arguments become the types in the type variable tuple:
+
+as this would make ``env`` a keyword-only argument.
+
+Unpacking an unbounded tuple is equivalent to the PEP 484 behavior
```
"Using an unpacked unbounded tuple for the type of ``*args`` is..." would be a bit clearer I think.

mrahtz reviewed (approved):
Lots of small comments, but basically looks good. Thanks, Pradeep, for getting this done so quickly!

gvanrossum commented on pep-0646.rst:
```diff
@@ -414,6 +354,133 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+Unpacking a Tuple Type
+----------------------
+
+We mentioned that a ``TypeVarTuple`` simply stands for a tuple of
```
Also, I'd omit "simply". When that word is used, as a reader I frequently find that the concept is all *but* simple. And in other cases it just sounds condescending.

gvanrossum commented on pep-0646.rst:
```diff
@@ -414,6 +354,133 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
     z = prefix_tuple(x=0, y=(True, 'a'))
     # Inferred type of z is Tuple[int, bool, str]
 
+Unpacking a Tuple Type
```
We're already inconsistent, e.g. line 207 uses singular.

gvanrossum replied:
I'm guessing I am now definitely on the side of plurals.

gvanrossum reviewed (commented):
I did not do a thorough review yet, but here are two nits.

In general, I worry a bit that this feels like a big rewrite, and as such the SC was right to un-approve the PEP and put it back in the queue. I doubt that they will end up rejecting it, but we are making more work for them, unless we add something like a cover letter when we resubmit saying "the update doesn't change anything fundamental, it just clarifies a bunch of stuff, and it makes *one* significant change (lifting the restriction on types with unknown length)." And they will still want to see for themselves that we're not (accidentally or intentionally) smuggling in something else.

pradeep90 commented on pep-0646.rst:
```diff
@@ -414,6 +354,133 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+zero or more elements of type ``str`` in between. Note that
+``Tuple[*Tuple[int, ...]]`` is equivalent to ``Tuple[int, ...]``.
+
+Consider the following function:
```
Agreed. In any case, the next two examples clarify what happens when we pass in an unbounded tuple, so I'll just remove this.

pradeep90 commented on pep-0646.rst:
```diff
@@ -414,6 +354,133 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+unpacking ``Tuple[bool, bool]`` in the return type ``Tuple[float, *Ts,
+float]`` gives back ``Tuple[float, bool, bool, float]``.
+
+Unpacking an Unbounded Tuple Type
```
I find the singular version clearer, so I'll keep it for now. Maybe take a look after everything is resolved; if you still feel it's better, I won't object.

gvanrossum replied:
Plural...

pradeep90 commented:
@mrahtz Addressed all the comments.

@gvanrossum 

> In general, I worry a bit that this feels like a big rewrite

Yeah, I ignored the political aspect of it and just described from a technical point of view how tuple types affected certain sections :) Now that it's written down, we can trim any parts you feel aren't worth adding.

> and as such the SC was right to un-approve the PEP and put it back in the queue. I doubt that they will end up rejecting it, but we are making more work for them, unless we add something like a cover letter when we resubmit saying "the update doesn't change anything fundamental, it just clarifies a bunch of stuff, and it makes one significant change (lifting the restriction on types with unknown length)." And they will still want to see for themselves that we're not (accidentally or intentionally) smuggling in something else.

A cover letter sounds good. We could say that we have preserved all of the old behavior and described some behavior that was previously unspecified or forbidden.

Main changes:

1. Describe how to unpack tuple types.
2. Remove section "Type Variable Tuples Must Have Known Length", which was an artificial restriction.
3. Explain how `Array` without type arguments actually works.
4. Explain how `*args: *Tuple[<...>]` works.
5. Explain how `Callable[[int, *Ts, str], None]` works.

We can cut (4) if needed. Readers can infer it using (1).

mrahtz commented:
@gvanrossum Just checking, are the next steps here for you to give us a thorough review of these changes, or for us to try making the delta smaller to minimise SC concerns? (If it's the former and you just haven't got round to it yet, no worries!)

gvanrossum commented on pep-0646.rst:
```diff
@@ -167,6 +170,9 @@ Type variable tuples are created with:
 
     Ts = TypeVarTuple('Ts')
 
+Using a Type Variable Tuple in Generic Classes
```
I know we already talked about plural vs. singular, but in this case (and the next) I find the use of singular "Type Variable Tuple" with plural "Generic Classes" or "Functions" jarring. I propose to use plurals all the way.

gvanrossum commented on pep-0646.rst:
```diff
@@ -414,6 +355,100 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+        x: Array[Batch, *Shape]
+    ) -> None: ...
+
+    expect_variadic_array(y)  # OK
```
Why is this okay? y's shape may have length 0 (the end case of `Tuple[Any, ...]`) but x needs to have at least one dimension (`Batch`).

pradeep90 replied:
`Tuple[X, ...]` represents a tuple of arbitrary length, not zero. For example, `xs: Tuple[int, ...]; xs[0]` is not a type error. It can be unsafe, yes, but that comes with the territory when we use `Tuple[int, ...]`.

This is what enables gradual typing of a large code base. A variable that hasn't yet been given shape types (`x: Array`) can be used without complaint wherever a precise shape type is expected (`y: Array[A, B, C]` or `y: Array[A, *Ts]`). I had added this example and the paragraph below to demonstrate this.

gvanrossum commented on pep-0646.rst:
```diff
@@ -414,6 +355,100 @@ Normal ``TypeVar`` instances can also be prefixed and/or suffixed:
+Multiple Unpackings in a Tuple: Not Allowed
+'''''''''''''''''''''''''''''''''''''''''''
+
+As with ``TypeVarTuple``, only one unpacking may appear in a tuple:
```
Explain why? (Maybe put the explanation in the corresponding section for TVT.) IIRC, the reason is that of we have e.g. [int, str, str, str, str], there are multiple solutions for Ts and Ts2.

gvanrossum commented on pep-0646.rst:
```diff
@@ -428,13 +463,59 @@ individual arguments become the types in the type variable tuple:
+
+    def execle(path: str, *args: *Ts, env: Env) -> None: ...
+
+as this would make ``env`` a keyword-only argument.
```
s/this/that/?

gvanrossum commented on pep-0646.rst:
```diff
@@ -506,6 +576,63 @@ to the type variable tuple:
 
     def foo(f: Callable[[int, *Ts, T], Tuple[T, *Ts]]): ...
 
+The behavior of a Callable containing an unpacked item, whether the
+item is a ``TypeVarTuple`` or a tuple type, is to treat the elements
+as if they were the type for ``*args``. So, ``Callable[[*Ts], None]``
```
Nice!

gvanrossum commented on pep-0646.rst:
```diff
@@ -506,6 +576,63 @@ to the type variable tuple:
+Behaviour when Type Parameters are not Specified
+------------------------------------------------
+
+When a generic class parameterised by a type variable tuple is used without
+any type parameters, it behaves as if the type variable tuple was
+substituted with ``Tuple[Any, ...]``:
```
Nice, isn't it! :-)

gvanrossum commented on pep-0646.rst:
```diff
@@ -654,8 +782,8 @@ otherwise imply. Also, we may later wish to support arguments that should not be
 
-Behaviour when Type Parameters are not Specified
-------------------------------------------------
+Unspecified Type Parameters: Tuples vs TypeVarTuples
```
```suggestion
Unspecified Type Parameters: Tuple vs TypeVarTuple
```
I may sound inconsistent, but the point in this case is that you're writing plurals of specific names, which makes me a bit uncomfortable. (If there was a type Child, would you write Children?)

gvanrossum reviewed (commented):
I'm pretty happy here! Bunch of plural/singular nits, and one place where I think an example should actually be an error (but maybe I'm missing something).

gvanrossum commented:
@pradeep90 Do you have time to finish this PR this week? We also have to consider #2177 and I'd like to get this moving to the SC again.

pradeep90 commented:
Done! Sorry, was working with Steven on the callable PEP. Lmk if you have further thoughts :)

@mrahtz We'll probably need to update the `*args: *Ts` grammar section to mention `*Tuple[<...>]`.

gvanrossum commented:
Aargh! I wish you wouldn't force-push. Now GH won't show "commits since last review". But no worries, I'll figure it out.

gvanrossum reviewed (approved):
LGTM. I will merge this now, then we can move on to the other PR for this PEP, and then we can send it back to the SC (you can start drafting a cover letter now).
